tensorboard/spanner.py

```diff
@@ -0,0 +1,514 @@
+
+  #def fetchmany(self, size=None):
+    #"""Returns next chunk of rows in result set.
+
+    #:type size: int
+    #"""
+    #self._check_that_read_query_was_issued()
+    #if size is not None:
+      #return self._delegate.fetchmany(size)
+    #else:
+      #return self._delegate.fetchmany()
+
+  #def fetchall(self):
+    #"""Returns next row in result set.
+
+    #:rtype: tuple[object]
+    #"""
+    #self._check_that_read_query_was_issued()
+    #return self._delegate.fetchone()
+
+  #@property
+  #def description(self):
+    #"""Returns information about each column in result set.
+
+    #See: https://www.python.org/dev/peps/pep-0249/
+
+    #:rtype: list[tuple[str, int, int, int, int, int, bool]]
+    #"""
+    #self._check_that_read_query_was_issued()
+    #return self._delegate.description
+
+  #@property
+  #def rowcount(self):
+    #"""Returns number of rows retrieved by last read query.
+
+    #:rtype: int
+    #"""
+    #self._check_that_read_query_was_issued()
+    #return self._delegate.rowcount
+
+  #@property
+  #def lastrowid(self):
+    #"""Returns last row ID.
+
+    #:rtype: int
+    #"""
+    #self._check_that_read_query_was_issued()
+    #return self._delegate.lastrowid
+
+  #def _get_arraysize(self):
+    #self._init_delegate()
+    #return self._delegate.arraysize
+
+  #def _set_arraysize(self, arraysize):
+    #self._init_delegate()
+    #self._delegate.arraysize = arraysize
+
+  #arraysize = property(_get_arraysize, _set_arraysize)
+
+  def close(self):
+    """Closes resources associated with cursor."""
+    # TODO(jlewi): Are there any spanner resources that should be released
+    pass
+
+  #def __iter__(self):
+    #"""Returns iterator over results of last read query.
+
+    #:rtype: types.GeneratorType[tuple[object]]
+    #"""
+    #self._check_that_read_query_was_issued()
+    #for row in self._delegate:
+      #yield row
+
+  #def nextset(self):
+    #"""Raises NotImplementedError."""
+    #raise NotImplementedError('Cursor.nextset not supported')
+
+  #def callproc(self, procname, parameters=()):
+    #"""Raises NotImplementedError."""
+    #raise NotImplementedError('Cursor.callproc not supported')
+
+  #def setinputsizes(self, sizes):
+    #"""Raises NotImplementedError."""
+    #raise NotImplementedError('Cursor.setinputsizes not supported')
+
+  #def setoutputsize(self, size, column):
+    #"""Raises NotImplementedError."""
+    #raise NotImplementedError('Cursor.setoutputsize not supported')
+
+  #def _init_delegate(self):
+    #self._check_closed()
+    #if self._delegate is None:
+      #self._delegate = self.connection._delegate.cursor()
+
+  #def _check_that_read_query_was_issued(self):
+    #self._check_closed()
+    #if self._delegate is None:
+      #raise ValueError('no read query was issued')
+
+  #def _check_closed(self):
+    #if self._is_closed:
+      #raise ValueError('cursor was closed')
+
+
+# TODO(jlewi): We need to override Schema because Cloud Spanner uses a slightly
```

Comment:
We're probably going to need to integrate some of the logic written in tensorboard.backend.application into tensorboard.db so we can have access to the `--db URI`. Then we can probably just have some if statements or something. I don't necessarily mind hard-coding the schema differences for all the different DB's into one place. It might be easier to reason about that way. What do you think?

Reply:
I'm leaning towards defining some Schema classes (TableSchema, ColumnSchema, etc...) that can be used to define the schemas in a DB independent way. Then we can just have DB specific ways of generating the corresponding DDL statements. I'll send you a PR in a couple days and we can iterate.